### PR TITLE
Shim all the d2i_* functions in OpenSSL.

### PR DESCRIPTION
**Pull request by eerhardt**
Adding shims for all the d2i functions.

@bartonjs, @nguerrera, @sokket

### CODE REVIEW

File: src/Native/System.Security.Cryptography.Native/pal_x509.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include "pal_x509.h"
+
+extern "C" X509_NAME* DecodeX509Name(const unsigned char* buf, int32_t len)
```
Comment (bartonjs) [inline src/Native/System.Security.Cryptography.Native/pal_x509.cpp]: Probably better belongs in pal_x509_name.cpp/h

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslPkcs12Reader.cs [hunk 2]
```diff
@@ -22,10 +22,7 @@ private OpenSslPkcs12Reader(SafePkcs12Handle pkcs12Handle)
 
         public unsafe static bool TryRead(byte[] data, out OpenSslPkcs12Reader pkcs12Reader)
```
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslPkcs12Reader.cs]: Do these methods that no longer need to call OpenSslD2I still need to be declared unsafe?

**Comment by bartonjs** [comment]
Minor feedback, otherwise LGTM.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslPkcs12Reader.cs (hunk 2 above)
Comment (eerhardt) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslPkcs12Reader.cs]: I've removed all the 'unsafe' declarations except one - it is still using pointers.

File: src/Native/System.Security.Cryptography.Native/pal_x509.cpp (hunk 1 above)
Comment (eerhardt) [inline src/Native/System.Security.Cryptography.Native/pal_x509.cpp]: I agree it logically does belong there, but following the [Interop Guidelines](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/interop-guidelines.md):

>- Minimize unnecessary IL in assemblies.
  - Interop signatures should only be compiled into the assemblies that
    actually consume them. Having extra signatures bloats assemblies and
    makes it more difficult to do static analysis over assemblies to
    understand what they actually use. It also leads to problems when such
    static verification is used as a gate, e.g. if a store verifies that
    only certain APIs are used by apps in the store.

If I was to put the signature in Interop.X509Name.cs, then System.Net.Security would have a new method compiled in it that it doesn't use, and System.Security.Cryptography.X509Certificates would have 2 new methods that it doesn't use.
(Note: I like to keep the Interop.X.cs file 1:1 with pal_x.cpp/h, as I think this makes the most sense when trying to find where methods are declared.)

**Comment by eerhardt** [comment]
@dotnet-bot test this please (Socket test failed, which I don't think I broke)

File: src/Native/System.Security.Cryptography.Native/pal_x509.cpp (hunk 1 above)
Comment (bartonjs) [inline src/Native/System.Security.Cryptography.Native/pal_x509.cpp]: X509Certificates will end up using the methods that are in X509Name as you make further shimming progress; and the SafeHandle should probably move into Interop.X509Name.cs as well (or have the file further exploded).

So my preferences would be
1. Move this to pal_x509_name, and Interop.X509Name.cs
2. Move this to pal_x509_name, move the extern to Interop.X509Name_Decode.cs or Interop.DecodeX509Name.cs
3. Move this to pal_x509_name, leave it in Interop.X509.cs
4. Leave the native method in pal_x509, the extern method in Interop.X509.cs
5. Leave the native method in pal_x509, move the extern to Interop.X509Name.cs

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslPkcs12Reader.cs (hunk 2 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslPkcs12Reader.cs]: Danke :)

**Comment by nguerrera** [comment]
LGTM.

**Comment by stephentoub** [comment]
@dotnet-bot test this please (#3494)

**Comment by stephentoub** [comment]
LGTM

**Comment by eerhardt** [comment]
FYI - I still have 1 comment to address from Jeremy, so please don't merge this yet.

**Comment by eerhardt** [comment]
All code review comments have been addressed and squashed.

**Pull request merged**